### PR TITLE
Add tests to verify corner cases of PI DMA

### PR DESCRIPTION
**Pull request by rasky**
This should cover most (known...) edge cases of PI DMA

### CODE REVIEW

File: src/tests/cart_memory/dma.rs [hunk 1]
```diff
@@ -1,15 +1,31 @@
...
 use crate::tests::{Level, Test};
-use crate::tests::soft_asserts::soft_assert_eq;
+use crate::tests::soft_asserts::{soft_assert_eq, soft_assert_eq2};
 use crate::uncached_memory::UncachedHeapMemory;
 
+const RDRAM_PAGE_SIZE: usize = 2048;
+
 const DATA: [u64; 4] = [0x01234567_89ABCDEF, 0x21436587_99BADCFE, 0xA9887766_55443322, 0x32445566_29384756];
 
+// This should really be an array of u8, but we need to use u16 to gurantee 2-byte alignment
```
Comment (lemmy-64) [inline src/tests/cart_memory/dma.rs:18]:
```suggestion
// This should really be an array of u8, but we need to use u16 to guarantee 2-byte alignment
```

File: src/tests/cart_memory/dma.rs [hunk 2]
```diff
@@ -1,15 +1,31 @@
...
+const COUNTER16: [u16; 512] = {
+    let mut result = [0u16; 512];
+    let mut i = 0u16;
+    while i < 512 {
+        result[(i/2) as usize] = (i & 255) * 256 + ((i & 255)+1);
```
Comment (lemmy-64) [inline src/tests/cart_memory/dma.rs:23]: my IDE formats this stuff...I should setup rustfmt so that contributions are properly formatted as well...

**Review by lemmy-64 — APPROVED** [review]
